File: android/app/src/main/java/com/testthing/myapp/CognitoPlugin.kt
package com.testthing.myapp

import com.amazonaws.mobileconnectors.cognitoauth.Auth
import com.amazonaws.mobileconnectors.cognitoauth.AuthUserSession
import com.amazonaws.mobileconnectors.cognitoauth.handlers.AuthHandler
import com.getcapacitor.JSObject
import com.getcapacitor.NativePlugin
import com.getcapacitor.Plugin
import com.getcapacitor.PluginCall
import com.getcapacitor.PluginMethod
//import com.testthing.myapp.cognito

@NativePlugin
class CognitoPlugin : Plugin() {

    private var jwt: String? = null
    private var signInInProgress = false
    private var getSessionPromise:Promise? = null
    private var cognitoAuth:Auth? = null

    @PluginMethod
    fun config(call: PluginCall) {
        val appClientId = call.getString("appClientId")
        val signIn = call.getString("signInRedirectUri")
        val signOut = call.getString("signOutRedirectUri")
        val webDomain = call.getString("webDomain")
        val scopes = arrayOf("email", "openid")
        val builder = Auth.Builder()
                .setApplicationContext(context)
                .setAppClientId(appClientId)
                .setAppCognitoWebDomain(webDomain)
                .setSignInRedirect(signIn)
                .setSignOutRedirect(signOut)
                .setScopes(scopes as MutableSet<String>?)

        builder.setAuthHandler(TheAuthHandler())
        cognitoAuth = builder.build()

        call.resolve()
    }

    @PluginMethod
    fun getSession(call: PluginCall) {
        call.resolve()
    }

    @PluginMethod
    fun isSignedIn(call: PluginCall) {
        call.resolve()
    }

    @PluginMethod
    fun signOut(call: PluginCall) {
        call.resolve()
    }

    internal inner class TheAuthHandler : AuthHandler {
        override fun onSuccess(session: AuthUserSession) {
            // This will invoked to return tokens on successful authentication or when valid tokens are available locally.
            // 'session' will contain valid tokens for the user.
//            Timber.e("Success")
            jwt = session.accessToken.jwtToken
            signInInProgress = false
            val map = Arguments.createMap()
            map.putString("session", jwt)
            // Resolve the global promise with the jwt
            getSessionPromise?.resolve(map)
            getSessionPromise = null
//            Timber.e("jwt: %s", jwt)
        }

        override fun onSignout() {
//            Timber.e("Signed out")
            signInInProgress = false
        }

        override fun onFailure(e: Exception) {
//            Timber.e("Failure Message: %s", e.localizedMessage)
//            Timber.e("Failure Stack Trace:")
            e.printStackTrace()
            getSessionPromise?.reject(e)
            getSessionPromise = null
        }

    }
}

File: ios/App/App/CognitoPlugin.swift
//
//  CognitoBridge.swift
//  App
//
//  Created by Weed, Chris on 2/8/19.
//

import AWSCognitoAuth
import Capacitor

@objc(CognitoPlugin)
public class CognitoPlugin: CAPPlugin, AWSCognitoAuthDelegate {
    public func getViewController() -> UIViewController {
        return self.bridge.viewController
    }
    
    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": value
            ])
    }
    
    @objc func config(_ call: CAPPluginCall) {
        let appClientId = call.getString("appClientId") ?? "";
        let scopes = Set<String>(["openid", "email"]); // TODO
        let signInRedirectUri = call.getString("signInRedirectUri") ?? "";
        let signOutRedirectUri = call.getString("signOutRedirectUri") ?? "";
        let webDomain = call.getString("webDomain") ?? "";
        
        let config = AWSCognitoAuthConfiguration(appClientId: appClientId, appClientSecret: nil, scopes: scopes, signInRedirectUri: signInRedirectUri, signOutRedirectUri: signOutRedirectUri, webDomain: webDomain);
        
        AWSCognitoAuth.registerCognitoAuth(with: config, forKey: "AWSCognito")
        call.resolve()
    }
    
    @objc func getSession(_ call: CAPPluginCall) {
        let cognitoAuth = AWSCognitoAuth(forKey: "AWSCognito");
        cognitoAuth.delegate = self
        
        cognitoAuth.getSession(self.bridge.viewController) { (session, err) in
            if(err != nil) {
                NSLog(err.debugDescription)
                call.reject(err.debugDescription)
            } else {
                call.resolve([
                    "accessToken": session?.accessToken?.tokenString ?? ""])
            }
        }
    }
    
    @objc func signOut(_ call: CAPPluginCall) {
        let cognitoAuth = AWSCognitoAuth(forKey: "AWSCognito");
        cognitoAuth.delegate = self
        
        cognitoAuth.signOut { (err) in
            if(err != nil) {
                NSLog(err.debugDescription);
                call.reject(err.debugDescription);
            } else {
                call.resolve()
            }
        }
    }
    
    @objc func isSignedIn(_ call: CAPPluginCall) {
        let cognitoAuth = AWSCognitoAuth(forKey: "AWSCognito");
        
        call.resolve(["isSignedIn": cognitoAuth.isSignedIn]);
    }
}
